File: android/src/main/java/com/thoagf/flutter_traffic_stats/networt_stats/NetworkStats.java
package com.thoagf.flutter_traffic_stats.networt_stats;

import android.annotation.SuppressLint;
import android.app.usage.NetworkStats.Bucket;
import android.app.usage.NetworkStatsManager;
import android.content.Context;
import android.content.pm.ApplicationInfo;
import android.net.NetworkCapabilities;
import android.os.Build;
import android.telephony.TelephonyManager;

import androidx.annotation.RequiresApi;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class NetworkStats {
    @SuppressLint({"MissingPermission", "HardwareIds"})
    private static String getSubscriberId(Context context) {
        try {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q)
                return null;
            TelephonyManager telephonyManager = (TelephonyManager) context.getSystemService(Context.TELEPHONY_SERVICE);
            return telephonyManager != null ? telephonyManager.getSubscriberId() : null;
        } catch (Exception ignored) {
        }
        return "";
    }

    @RequiresApi(api = Build.VERSION_CODES.M)
    public static List<Map<String, Object>> queryNetworkBuckets(Context context, long startDate, long endDate, int type) {
        NetworkType networkType = NetworkType.All;
        if(type==2){
            networkType = NetworkType.WiFi;
        }else if(type==3){
            networkType = NetworkType.Mobile;
        }

        NetworkStatsManager networkStatsManager = (NetworkStatsManager) context.getSystemService(Context.NETWORK_STATS_SERVICE);
        String subscriberID = getSubscriberId(context);
        List<ApplicationInfo> installedApplications = context.getPackageManager().getInstalledApplications(0);
        List<Map<String, Object>> result = new ArrayList<>();
        for (ApplicationInfo appInfo : installedApplications) {
            List<Bucket> buckets = getNetworkBuckets(appInfo, networkType, networkStatsManager, startDate, endDate, subscriberID);
            Map<String, Object> appData = new HashMap<>();
            appData.put("packageName", appInfo.packageName);
            List<Map<String, Object>> bucketData = new ArrayList<>();
            for (Bucket bucket : buckets) {
                Map<String, Object> bucketMap = new HashMap<>();
                bucketMap.put("rxBytes", bucket.getRxBytes());
                bucketMap.put("rxPackets", bucket.getRxPackets());
                bucketMap.put("txBytes", bucket.getTxBytes());
                bucketMap.put("txPackets", bucket.getTxPackets());
                bucketMap.put("startTimeStamp", bucket.getStartTimeStamp());
                bucketMap.put("endTimeStamp", bucket.getEndTimeStamp());
                bucketData.add(bucketMap);
            }
            appData.put("data", bucketData);
            result.add(appData);
        }
        return result;
    }

    @SuppressLint("MissingPermission")
    @RequiresApi(api = Build.VERSION_CODES.M)
    private static List<Bucket> getNetworkBuckets(ApplicationInfo appInfo, NetworkType networkType, NetworkStatsManager networkStatsManager, long startDate, long endDate, String subscriberID) {
        switch (networkType) {
            case Mobile:
                return getNetworkBuckets(NetworkCapabilities.TRANSPORT_CELLULAR, networkStatsManager, startDate, endDate, subscriberID, appInfo);
            case WiFi:
                return getNetworkBuckets(NetworkCapabilities.TRANSPORT_WIFI, networkStatsManager, startDate, endDate, subscriberID,appInfo);
            default:
                return new ArrayList<>();
        }
    }

    @SuppressLint("MissingPermission")
    @RequiresApi(api = Build.VERSION_CODES.M)
    private static List<Bucket> getNetworkBuckets(int networkType, NetworkStatsManager networkStatsManager, long startDate, long endDate, String subscriberID, ApplicationInfo info) {
        try {
            List<Bucket> result = new ArrayList<>();
            android.app.usage.NetworkStats queryDetailsForUid = networkStatsManager.queryDetailsForUid(networkType, subscriberID, startDate, endDate, info.uid);
            while (queryDetailsForUid.hasNextBucket()) {
                Bucket tmpBucket = new Bucket();
                queryDetailsForUid.getNextBucket(tmpBucket);
                result.add(tmpBucket);
            }
            return result;
        } catch (Exception ignored) {
            return new ArrayList<>();
        }
    }

    private enum NetworkType {
        All,
        WiFi,
        Mobile,
    }
}

File: android/src/main/java/com/thoagf/flutter_traffic_stats/networt_stats/NetworkStats.kt
package com.thoagf.flutter_traffic_stats.networt_stats

import android.annotation.SuppressLint
import android.app.usage.NetworkStats
import android.app.usage.NetworkStats.Bucket
import android.app.usage.NetworkStatsManager
import android.content.Context
import android.content.pm.ApplicationInfo
import android.net.NetworkCapabilities
import android.os.Build
import android.telephony.TelephonyManager
import android.util.Log
import androidx.annotation.RequiresApi


object NetworkStats {

    @SuppressLint("MissingPermission", "HardwareIds")
    private fun getSubscriberId(
        context: Context,
    ): String? {
        try {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q)
                return null

            val telephonyManager = context.getSystemService(Context.TELEPHONY_SERVICE) as TelephonyManager?
            return telephonyManager?.subscriberId
        } catch (_: Exception) {
        }

        return ""
    }

    @RequiresApi(Build.VERSION_CODES.M)
    fun queryNetworkBuckets(context: Context, startDate: Long, endDate: Long, type: Int): List<Map<String, Any>> {
        val networkType: NetworkType = when (type) {
            1 -> NetworkType.All
            2 -> NetworkType.WiFi
            3 -> NetworkType.Mobile
            else -> NetworkType.All
        }

        val networkStatsManager =
            context.getSystemService(Context.NETWORK_STATS_SERVICE) as NetworkStatsManager
        val subscriberID = getSubscriberId(context)

        val installedApplications: MutableList<ApplicationInfo> =
            context.packageManager.getInstalledApplications(0)

        return installedApplications.map { appInfo: ApplicationInfo ->

            val buckets = appInfo.getNetworkBuckets(
                networkStatsManager = networkStatsManager,
                startDate = startDate,
                endDate = endDate,
                networkType = networkType,
                subscriberID = subscriberID,
            )
            mapOf(
                "packageName" to appInfo.packageName,
                "data" to buckets.map {
                    bucket -> mapOf(
                        "rxBytes" to bucket.rxBytes,
                        "rxPackets" to bucket.rxPackets,
                        "txBytes" to bucket.txBytes,
                        "txPackets" to bucket.txPackets,
                        "startTimeStamp" to bucket.startTimeStamp,
                        "endTimeStamp" to bucket.endTimeStamp,
                    )
                }
            )
        }
    }

    @SuppressLint("MissingPermission")
    @RequiresApi(Build.VERSION_CODES.M)
    private fun ApplicationInfo.getNetworkBuckets(
        networkType: NetworkType,
        networkStatsManager: NetworkStatsManager,
        startDate: Long,
        endDate: Long,
        subscriberID: String? = null,
    ): List<Bucket> {
        when(networkType){
            NetworkType.Mobile -> {
                return getNetworkBuckets(
                    NetworkCapabilities.TRANSPORT_CELLULAR,
                    networkStatsManager, startDate, endDate, subscriberID
                )
            }
            NetworkType.WiFi -> {
                return getNetworkBuckets(
                    NetworkCapabilities.TRANSPORT_WIFI,
                    networkStatsManager, startDate, endDate, subscriberID
                )
            }
            else -> {
//                var mobileStats = getNetworkBuckets(
//                    NetworkCapabilities.TRANSPORT_CELLULAR,
//                    networkStatsManager, startDate, endDate
//                )
//                var wifiStats = getNetworkBuckets(
//                    NetworkCapabilities.TRANSPORT_WIFI,
//                    networkStatsManager, startDate, endDate
//                )
                return emptyList()
            }
        }
    }

    @SuppressLint("MissingPermission")
    @RequiresApi(Build.VERSION_CODES.M)
    private fun ApplicationInfo.getNetworkBuckets(
        networkType: Int,
        networkStatsManager: NetworkStatsManager,
        startDate: Long,
        endDate: Long,
        subscriberID: String? = null,
    ): List<Bucket> {
        return try {
            val result = ArrayList<Bucket>()
            val queryDetailsForUid: NetworkStats = networkStatsManager.queryDetailsForUid(
                networkType, subscriberID, startDate, endDate, uid
            )
            while (queryDetailsForUid.hasNextBucket()) {
                val tmpBucket = Bucket()
                queryDetailsForUid.getNextBucket(tmpBucket)
                result.add(tmpBucket)
            }
            result
        } catch (err: Exception) {
            emptyList()
        }
    }


    private enum class NetworkType {
        All,
        WiFi,
        Mobile,
    }
}
